libevmasm/SemanticInformation.cpp

```diff
@@ -232,6 +232,19 @@ bool SemanticInformation::movable(Instruction _instruction)
 	return true;
 }
 
+bool SemanticInformation::movableIfStorageInvariant(Instruction _instruction)
+{
+	switch (_instruction)
+	{
+	case Instruction::EXTCODESIZE:
+	case Instruction::EXTCODEHASH:
+	case Instruction::SLOAD:
```

Comment:
Should also include:
 - balance
 - selfbalance
 - returndatasize (I think)

I think we could even add `staticcall` as long as the loop is both memory and state invariant. Maybe this should be added as an enum flag?

Reply:
staticcall itself invalidates memory, so in order to do this properly, we would have to exclude the item being propagated from the checks for side-effects - so maybe this is too complicated to do.